In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
data_path = '/content/drive/MyDrive/DeepfakeData/SampleData/organized'

In [3]:
import os

destination_base = '/content/drive/MyDrive/DeepfakeData/SampleData/frames'
subdirectories = [
    'train/real_videos',
    'train/fake_videos',
    'val/real_videos',
    'val/fake_videos',
    'test/real_videos',
    'test/fake_videos'
]

for subdir in subdirectories:
    full_path = os.path.join(destination_base, subdir)
    os.makedirs(full_path, exist_ok=True)

print("Directories created successfully.")

Directories created successfully.


In [4]:
import os

video_files = []
main_subdirs = ['train', 'val', 'test']
inner_subdirs = ['real_videos', 'fake_videos']

for main_subdir in main_subdirs:
    main_subdir_path = os.path.join(data_path, main_subdir)
    if os.path.isdir(main_subdir_path):
        for inner_subdir in inner_subdirs:
            inner_subdir_path = os.path.join(main_subdir_path, inner_subdir)
            if os.path.isdir(inner_subdir_path):
                for filename in os.listdir(inner_subdir_path):
                    if filename.endswith('.mp4'):
                        video_files.append(os.path.join(inner_subdir_path, filename))

print(f"Found {len(video_files)} video files.")

Found 400 video files.


In [ ]:
# import subprocess
# import os

# destination_base = '/content/drive/MyDrive/DeepfakeData/SampleData/frames'

# for video_file in video_files:
#     relative_path = os.path.relpath(video_file, data_path)
#     output_dir = os.path.join(destination_base, os.path.dirname(relative_path), os.path.splitext(os.path.basename(video_file))[0])
#     os.makedirs(output_dir, exist_ok=True)
#     output_frame_pattern = os.path.join(output_dir, '%04d.png')

#     ffmpeg_command = [
#         'ffmpeg',
#         '-i', video_file,
#         '-vf', 'fps=3',
#         output_frame_pattern
#     ]

#     try:
#         subprocess.run(ffmpeg_command, check=True, capture_output=True, text=True)
#         print(f"Successfully extracted frames from {video_file}")
#     except subprocess.CalledProcessError as e:
#         print(f"Error extracting frames from {video_file}: {e}")
#         print(f"Stderr: {e.stderr}")
#     except FileNotFoundError:
#         print("Error: ffmpeg command not found. Make sure ffmpeg is installed and in your PATH.")

Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/aelfnikyqj.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/afoovlsmtx.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/agrmhtjdlk.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/ajqslcypsw.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/anpuvshzoo.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/asaxgevnnp.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/atkdltyyen.mp4
Successfully extracted frames from /content/drive/MyDrive/DeepfakeData/SampleData/organized/train/real_videos/atvmxvwyns.mp4


In [ ]:
# #!/usr/bin/env python3
# """
# crop_faces.py — Detect, 1.3x enlarge, and save 299x299 face crops from frame folders.
# Useful to prepare ImageFolder datasets for training.

# Input frames dir structure:
# frames/
#   videoA/frame_00001.jpg
#   videoA/frame_00002.jpg
#   ...
#   videoB/frame_00001.jpg
#   ...

# Usage:
# Run the crop_faces_from_video_folder function with appropriate arguments.
# """

# import argparse, os
# from pathlib import Path
# import cv2
# import mediapipe as mp

# def enlarge_box(x1,y1,x2,y2, scale, W, H):
#     cx, cy = (x1+x2)/2.0, (y1+y2)/2.0
#     w, h = (x2-x1)*scale, (y2-y1)*scale
#     nx1, ny1 = max(0, int(cx - w/2)), max(0, int(cy - h/2))
#     nx2, ny2 = min(W-1, int(cx + w/2)), min(H-1, int(cy + h/2))
#     return nx1, ny1, nx2, ny2

# def detect_largest_face_mediapipe(img, face_detector):
#     # Convert the image to RGB as mediapipe requires RGB input
#     img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
#     results = face_detector.process(img_rgb)

#     if not results.detections:
#         return None

#     largest_face = None
#     largest_area = 0

#     for detection in results.detections:
#         bbox = detection.location_data.relative_bounding_box
#         ih, iw, _ = img.shape
#         x, y, w, h = int(bbox.xmin * iw), int(bbox.ymin * ih), int(bbox.width * iw), int(bbox.height * ih)
#         area = w * h
#         if area > largest_area:
#             largest_area = area
#             largest_face = (x, y, x+w, y+h)

#     return largest_face


# def crop_faces_from_video_folder(frames_dir, out_dir, enlarge_scale=1.3, jpeg_quality=95):
#     """
#     Detect, enlarge, and save face crops from frame folders.

#     Args:
#         frames_dir (str): Directory with per-video subfolders of frames.
#         out_dir (str): Where to write 299x299 crops.
#         enlarge_scale (float, optional): Scale to enlarge the face bounding box. Defaults to 1.3.
#         jpeg_quality (int, optional): JPEG quality for saved images. Defaults to 95.
#     """
#     mp_face_detection = mp.solutions.face_detection
#     mp_drawing = mp.solutions.drawing_utils

#     frames_root = Path(frames_dir)
#     out_root = Path(out_dir)
#     out_root.mkdir(parents=True, exist_ok=True)

#     videos = sorted([p for p in frames_root.iterdir() if p.is_dir()])
#     count = 0

#     # Initialize Mediapipe Face Detection
#     with mp_face_detection.FaceDetection(
#         model_selection=1, min_detection_confidence=0.5) as face_detector:

#         for vdir in videos:
#             # Change glob pattern to look for .png files
#             for fp in sorted(vdir.glob("*.png")):
#                 img = cv2.imread(str(fp))
#                 if img is None:
#                     continue
#                 H, W = img.shape[:2]
#                 box = detect_largest_face_mediapipe(img, face_detector)
#                 if box is None:
#                     continue
#                 x1,y1,x2,y2 = box
#                 x1,y1,x2,y2 = enlarge_box(x1,y1,x2,y2, enlarge_scale, W, H)
#                 crop = img[y1:y2, x1:x2]
#                 crop = cv2.resize(crop, (299,299), interpolation=cv2.INTER_AREA)
#                 out_fp = out_root / f"{vdir.name}_{fp.stem}.jpg" # Save as .jpg as requested
#                 cv2.imwrite(str(out_fp), crop, [int(cv2.IMWRITE_JPEG_QUALITY), jpeg_quality])
#                 count += 1

#     print(f"Saved {count} crops to {out_root}")

# # Example usage (you can call this function in a separate cell):
# # frames_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/frames' # Replace with your frames directory
# # out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops/train/real' # Replace with your desired output directory
# # crop_faces_from_video_folder(frames_dir, out_dir)

In [5]:
import os

frames_base_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/frames'
subdirectories = [
    'train/real_videos',
    'train/fake_videos',
    'val/real_videos',
    'val/fake_videos',
    'test/real_videos',
    'test/fake_videos'
]

for subdir in subdirectories:
    frames_subdir = os.path.join(frames_base_dir, subdir)
    print(f"Checking directory: {frames_subdir}")
    if os.path.isdir(frames_subdir):
        print("Directory exists.")
        # Check for files within the directory (limit to first few for brevity)
        files_in_dir = os.listdir(frames_subdir)
        if files_in_dir:
            print(f"Found {len(files_in_dir)} items in directory. First 5: {files_in_dir[:5]}")
            # Further check if any expected image files exist within the subdirectories (video folders)
            video_subdirs = [d for d in os.listdir(frames_subdir) if os.path.isdir(os.path.join(frames_subdir, d))]
            if video_subdirs:
                print(f"Found {len(video_subdirs)} video subdirectories. Checking the first one...")
                first_video_subdir = video_subdirs[0]
                first_video_subdir_path = os.path.join(frames_subdir, first_video_subdir)
                frames_in_video_subdir = [f for f in os.listdir(first_video_subdir_path) if f.endswith('.jpg') or f.endswith('.png')]
                if frames_in_video_subdir:
                    print(f"Found {len(frames_in_video_subdir)} image files in {first_video_subdir}. First 5: {frames_in_video_subdir[:5]}")
                else:
                    print(f"No image files found in {first_video_subdir}.")
            else:
                print("No video subdirectories found within this directory.")
        else:
            print("Directory is empty.")
    else:
        print("Directory does not exist.")
    print("-" * 30)

Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/frames/train/real_videos
Directory exists.
Found 61 items in directory. First 5: ['aelfnikyqj', 'afoovlsmtx', 'agrmhtjdlk', 'ajqslcypsw', 'anpuvshzoo']
Found 61 video subdirectories. Checking the first one...
Found 30 image files in aelfnikyqj. First 5: ['0001.png', '0002.png', '0003.png', '0004.png', '0005.png']
------------------------------
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/frames/train/fake_videos
Directory exists.
Found 258 items in directory. First 5: ['aagfhgtpmv', 'aapnvogymq', 'abofeumbvv', 'acifjvzvpm', 'aczrgyricp']
Found 258 video subdirectories. Checking the first one...
Found 30 image files in aagfhgtpmv. First 5: ['0001.png', '0002.png', '0003.png', '0004.png', '0005.png']
------------------------------
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/frames/val/real_videos
Directory exists.
Found 8 items in directory. First 5: ['abarnvbtwb', 'avmjormvsx'

In [ ]:
# import os

# # Define the base directories
# frames_base_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/frames'
# crops_base_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops'

# # Define the subdirectories for frames and crops
# subdirectories = [
#     'train/real_videos',
#     'train/fake_videos',
#     'val/real_videos',
#     'val/fake_videos',
#     'test/real_videos',
#     'test/fake_videos'
# ]

# # Process each subdirectory
# for subdir in subdirectories:
#     frames_subdir = os.path.join(frames_base_dir, subdir)
#     crops_subdir = os.path.join(crops_base_dir, subdir)

#     # Ensure the output directory exists
#     os.makedirs(crops_subdir, exist_ok=True)

#     print(f"Processing frames from: {frames_subdir}")
#     print(f"Saving crops to: {crops_subdir}")

#     # Call the crop_faces_from_video_folder function
#     crop_faces_from_video_folder(frames_subdir, crops_subdir)
#     print("-" * 30)

# print("Face cropping process completed for all subdirectories.")

Processing frames from: /content/drive/MyDrive/DeepfakeData/SampleData/frames/train/real_videos
Saving crops to: /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/real_videos
Saved 1789 crops to /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/real_videos
------------------------------
Processing frames from: /content/drive/MyDrive/DeepfakeData/SampleData/frames/train/fake_videos
Saving crops to: /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/fake_videos
Saved 7387 crops to /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/fake_videos
------------------------------
Processing frames from: /content/drive/MyDrive/DeepfakeData/SampleData/frames/val/real_videos
Saving crops to: /content/drive/MyDrive/DeepfakeData/SampleData/crops/val/real_videos
Saved 240 crops to /content/drive/MyDrive/DeepfakeData/SampleData/crops/val/real_videos
------------------------------
Processing frames from: /content/drive/MyDrive/DeepfakeData/SampleData/frames/val/fake_vi

In [6]:
import os

crops_base_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops'
subdirectories = [
    'train/real_videos',
    'train/fake_videos',
    'val/real_videos',
    'val/fake_videos',
    'test/real_videos',
    'test/fake_videos'
]

print("Checking if crop directories contain data:")

for subdir in subdirectories:
    crops_subdir = os.path.join(crops_base_dir, subdir)
    print(f"Checking directory: {crops_subdir}")
    if os.path.isdir(crops_subdir):
        files_in_dir = os.listdir(crops_subdir)
        if files_in_dir:
            print(f"Directory exists and contains {len(files_in_dir)} files. First 5: {files_in_dir[:5]}")
        else:
            print("Directory exists but is empty.")
    else:
        print("Directory does not exist.")
    print("-" * 30)

Checking if crop directories contain data:
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/real_videos
Directory exists and contains 1789 files. First 5: ['cfxkpiweqt_0015.jpg', 'cfxkpiweqt_0016.jpg', 'cfxkpiweqt_0017.jpg', 'cfxkpiweqt_0018.jpg', 'cfxkpiweqt_0019.jpg']
------------------------------
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/crops/train/fake_videos
Directory exists and contains 7387 files. First 5: ['ecuvtoltue_0029.jpg', 'ecuvtoltue_0030.jpg', 'ecwaxgutkc_0001.jpg', 'ecwaxgutkc_0002.jpg', 'ecwaxgutkc_0003.jpg']
------------------------------
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/crops/val/real_videos
Directory exists and contains 240 files. First 5: ['abarnvbtwb_0001.jpg', 'abarnvbtwb_0002.jpg', 'abarnvbtwb_0003.jpg', 'abarnvbtwb_0004.jpg', 'abarnvbtwb_0005.jpg']
------------------------------
Checking directory: /content/drive/MyDrive/DeepfakeData/SampleData/crops/val/fake_videos
Dire

In [7]:
%pip install timm==0.6.13

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 549.1/549.1 kB 10.5 MB/s eta 0:00:00
  Attempting uninstall: timm
    Found existing installation: timm 1.0.22
    Uninstalling timm-1.0.22:
      Successfully uninstalled timm-1.0.22


In [8]:
!pip install torch_xla

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.3/80.3 MB 31.3 MB/s eta 0:00:00


In [9]:
!pip install --upgrade torchvision

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 69.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 899.7/899.7 MB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 594.3/594.3 MB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 155.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 MB 30.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 954.8/954.8 kB 68.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.1/193.1 MB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 82.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.6/63.6 MB 40.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.5/267.5 MB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 288.2/288.2 MB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.3/39.3 MB 68.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.0

In [7]:
#!/usr/bin/env python3
"""
train_xception.py — Fine-tune XceptionNet (via timm) on face crops (real/fake)
Dataset layout (ImageFolder style):
crops/
  train/real/*.jpg
  train/fake/*.jpg
  val/real/*.jpg
  val/fake/*.jpg

Usage:
Run the train_model function with appropriate arguments.
"""

import argparse, os, random, time, math
from pathlib import Path

import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

# Import torch_xla
# import torch_xla
# import torch_xla.core.xla_model as xm

try:
    import timm
except Exception as e:
    raise SystemExit("Please install timm: pip install timm==1.*") from e

from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, average_precision_score, log_loss

IMAGENET_MEAN = [0.485, 0.456, 0.406]
IMAGENET_STD  = [0.229, 0.224, 0.225]

In [8]:
def seed_everything(seed: int = 42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

def build_loaders(data_dir, img_size=299, batch_size=64, num_workers=8):
    train_tf = transforms.Compose([
        transforms.RandomResizedCrop(img_size, scale=(0.9, 1.0)),
        transforms.RandomApply([transforms.GaussianBlur(3)], p=0.2),
        transforms.ColorJitter(0.1,0.1,0.1,0.05),
        transforms.ToTensor(),
        transforms.Normalize(IMAGENET_MEAN, IMAGENET_STD),
    ])
    val_tf = transforms.Compose([
        transforms.Resize(img_size),
        transforms.CenterCrop(img_size),
        transforms.ToTensor(),
        transforms.Normalize(IMAGENET_MEAN, IMAGENET_STD),
    ])
    train_ds = datasets.ImageFolder(os.path.join(data_dir, "train"), transform=train_tf)
    val_ds   = datasets.ImageFolder(os.path.join(data_dir, "val"),   transform=val_tf)
    train_loader = DataLoader(train_ds, batch_size=batch_size, shuffle=True,
                              num_workers=num_workers, pin_memory=True)
    val_loader   = DataLoader(val_ds,   batch_size=batch_size, shuffle=False,
                              num_workers=num_workers, pin_memory=True)
    return train_loader, val_loader, train_ds.classes

In [9]:
def evaluate_frame_level(model, loader, device, alpha=1.0, threshold=0.5):
    """
    Evaluate model and compute metrics including weighted precision.

    Args:
        model: The model to evaluate
        loader: DataLoader for evaluation
        device: Device to run evaluation on
        alpha: Weight factor for false positives in weighted precision.
               alpha = (ratio of real:fake in organic traffic) / (ratio of real:fake in dataset)
               For example, if organic is 1M:1 and dataset is 1:1, alpha = 1M
               Default 1.0 means no weighting (standard precision)
        threshold: Classification threshold for fake class (default 0.5)

    Returns:
        Dictionary of metrics including weighted precision (wP)
    """
    from sklearn.metrics import confusion_matrix

    model.eval()
    all_logits, all_targets = [], []
    with torch.no_grad():
        for x, y in loader:
            x = x.to(device)
            logits = model(x)
            all_logits.append(logits.cpu())
            all_targets.append(y)
    logits = torch.cat(all_logits, dim=0)
    targets = torch.cat(all_targets, dim=0).numpy()
    probs = torch.softmax(logits, dim=1).numpy()[:,1]
    preds = (probs >= threshold).astype(np.int64)

    # Calculate confusion matrix
    # Assuming: 0 = real, 1 = fake
    cm = confusion_matrix(targets, preds)
    if cm.shape == (2, 2):
        TN, FP, FN, TP = cm.ravel()
    else:
        # Handle edge cases where one class is missing
        if len(np.unique(targets)) == 1:
            if targets[0] == 0:  # All real
                TN, FP, FN, TP = len(targets), 0, 0, 0
            else:  # All fake
                TN, FP, FN, TP = 0, 0, 0, len(targets)
        else:
            TN, FP, FN, TP = 0, 0, 0, 0

    metrics = {}
    metrics["acc"] = float(accuracy_score(targets, preds))
    metrics["f1"]  = float(f1_score(targets, preds))

    # Calculate standard precision and recall
    if TP + FP > 0:
        metrics["precision"] = float(TP / (TP + FP))
    else:
        metrics["precision"] = 0.0

    if TP + FN > 0:
        metrics["recall"] = float(TP / (TP + FN))
    else:
        metrics["recall"] = 0.0

    # Calculate weighted precision: wP = TP / (TP + α * FP)
    if TP + alpha * FP > 0:
        metrics["wP"] = float(TP / (TP + alpha * FP))
    else:
        metrics["wP"] = 0.0

    # Additional metrics
    try:
        metrics["auc"] = float(roc_auc_score(targets, probs))
        metrics["ap_fake"] = float(average_precision_score(targets, probs))
        # Calculate log loss (binary cross-entropy)
        # Clip probabilities to avoid log(0) issues
        probs_clipped = np.clip(probs, 1e-15, 1 - 1e-15)
        metrics["log_loss"] = float(log_loss(targets, probs_clipped))
    except Exception:
        metrics["auc"] = float("nan")
        metrics["ap_fake"] = float("nan")
        metrics["log_loss"] = float("nan")

    # Store confusion matrix components for debugging
    metrics["TP"] = int(TP)
    metrics["FP"] = int(FP)
    metrics["FN"] = int(FN)
    metrics["TN"] = int(TN)

    return metrics

In [10]:
def save_checkpoint(state, is_best, out_dir, name="xception"):
    out_dir = Path(out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    ckpt = out_dir / f"{name}_last_dec2.pt"
    torch.save(state, ckpt)
    if is_best:
        best = out_dir / f"{name}_best_dec2.pt"
        torch.save(state, best)

In [12]:
def train_model(data_dir, out_dir, epochs_head=3, epochs_full=15, batch_size=64, img_size=299, num_workers=8, seed=42, lr_head=3e-4, lr_full=1e-4, weight_decay=1e-4, resume=True, alpha=1.0, use_class_weights=True):
    """
    Train Xception model with checkpoint resuming and weighted precision for model selection.

    Args:
        alpha: Weight factor for false positives in weighted precision.
               alpha = (ratio of real:fake in organic traffic) / (ratio of real:fake in dataset)
               For example, if organic traffic has 1M real:1 fake and dataset has 1:1, alpha = 1,000,000
               Default 1.0 means standard precision (no weighting).
               Higher alpha penalizes false positives more, which is important for class imbalance.
        use_class_weights: If True, use class weights in loss to penalize false positives more.
                          Weights are calculated automatically from dataset imbalance.
    """
    seed_everything(seed)

    # Use CUDA device if available, otherwise CPU
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")
    print(f"Using weighted precision with alpha={alpha} for model selection")
    print(f"  (alpha = ratio of real:fake in organic traffic / ratio of real:fake in dataset)")
    print(f"  (higher alpha penalizes false positives more, important for class imbalance)")

    train_loader, val_loader, classes = build_loaders(data_dir, img_size, batch_size, num_workers)
    print(f"Classes: {classes} (expect ['fake', 'real'] or similar two classes)")

    # Calculate class weights if requested (to penalize false positives more)
    class_weights = None
    if use_class_weights:
        # Count samples per class in training set
        from collections import Counter
        all_labels = []
        for _, labels in train_loader:
            all_labels.extend(labels.numpy())
        label_counts = Counter(all_labels)

        # Calculate weights: inverse frequency, normalized
        # Higher weight for minority class (real) to reduce false positives
        total_samples = sum(label_counts.values())
        num_classes = len(label_counts)
        class_weights = torch.FloatTensor([
            total_samples / (num_classes * label_counts[i]) if i in label_counts else 1.0
            for i in range(num_classes)
        ]).to(device)

        print(f"Class weights: {class_weights.cpu().numpy()} (higher weight = more penalty for misclassification)")
        print(f"  This will penalize false positives (real->fake) more during training")
    else:
        print("Not using class weights (standard CrossEntropyLoss)")

    # Check for existing checkpoint
    out_dir_path = Path(out_dir)
    checkpoint_path = out_dir_path / "xception_last.pt"
    start_epoch = 0
    best_val = -1.0
    resume_from_phase2 = False
    checkpoint = None

    if resume and checkpoint_path.exists():
        print(f"Found checkpoint at {checkpoint_path}. Loading...")
        checkpoint = torch.load(checkpoint_path, map_location=device)

        # Build model
        model = timm.create_model("xception", pretrained=False, num_classes=2)
        model.load_state_dict(checkpoint['model_state'])
        model = model.to(device)

        # Restore training state
        start_epoch = checkpoint.get('epoch', 0)
        # Restore best weighted precision, fallback to accuracy for old checkpoints
        val_metrics_ckpt = checkpoint.get('val_metrics', {})
        best_val = val_metrics_ckpt.get('wP', val_metrics_ckpt.get('acc', -1.0))

        # Check which phase we were in
        if start_epoch > epochs_head:
            # We were in Phase 2
            resume_from_phase2 = True
            print(f"Resuming from Phase 2, epoch {start_epoch - epochs_head}/{epochs_full} (total epoch {start_epoch})")
        else:
            # We were in Phase 1
            print(f"Resuming from Phase 1, epoch {start_epoch}/{epochs_head}")

        # Restore args if they exist (for validation)
        saved_args = checkpoint.get('args', {})
        if saved_args:
            print(f"Checkpoint was saved with args: epochs_head={saved_args.get('epochs_head')}, epochs_full={saved_args.get('epochs_full')}")
    else:
        # Build model from scratch
        print("No checkpoint found. Starting training from scratch with pretrained ImageNet weights.")
        model = timm.create_model("xception", pretrained=True, num_classes=2)
        model = model.to(device)

    # Phase 1: freeze backbone, train head
    head_params = []
    for n, p in model.named_parameters():
        if "fc" in n or "classifier" in n:
            head_params.append(p)
            p.requires_grad = True
        else:
            p.requires_grad = False

    opt = torch.optim.AdamW(head_params, lr=lr_head, weight_decay=weight_decay)
    crit = nn.CrossEntropyLoss(weight=class_weights) if class_weights is not None else nn.CrossEntropyLoss()

    # Load optimizer state if resuming from Phase 1
    if resume and checkpoint is not None and not resume_from_phase2:
        if 'optimizer_state' in checkpoint:
            opt.load_state_dict(checkpoint['optimizer_state'])
            print("Loaded optimizer state from checkpoint.")

    def train_one_epoch(epoch, model, loader, optimizer, criterion, device):
        model.train()
        losses = []
        for x, y in loader:
            x, y = x.to(device), y.to(device)
            optimizer.zero_grad(set_to_none=True)
            logits = model(x)
            loss = criterion(logits, y)
            loss.backward()
            optimizer.step()
            losses.append(loss.item())
        return float(np.mean(losses))

    # Phase 1: freeze backbone, train head (skip if resuming from Phase 2)
    if not resume_from_phase2:
        print("==> Phase 1: training the head")
        phase1_start = start_epoch + 1 if start_epoch < epochs_head else epochs_head + 1
        for epoch in range(phase1_start, epochs_head+1):
            tr_loss = train_one_epoch(epoch, model, train_loader, opt, crit, device)
            val_metrics = evaluate_frame_level(model, val_loader, device, alpha=alpha)
            # Use weighted precision for model selection (better for class imbalance)
            score = val_metrics["wP"]
            is_best = score > best_val
            if is_best: best_val = score
            save_checkpoint({
                "epoch": epoch,
                "model_state": model.state_dict(),
                "optimizer_state": opt.state_dict(),
                "val_metrics": val_metrics,
                "args": {
                    "data_dir": data_dir, "out_dir": out_dir, "epochs_head": epochs_head,
                    "epochs_full": epochs_full, "batch_size": batch_size, "img_size": img_size,
                    "num_workers": num_workers, "seed": seed, "lr_head": lr_head,
                    "lr_full": lr_full, "weight_decay": weight_decay, "alpha": alpha
                },
            }, is_best, out_dir, name="xception")
            print(f"[Head {epoch}/{epochs_head}] loss={tr_loss:.4f}  wP={score:.4f}  acc={val_metrics['acc']:.4f}  prec={val_metrics.get('precision', 0):.4f}  rec={val_metrics.get('recall', 0):.4f}  log_loss={val_metrics.get('log_loss', float('nan')):.4f}  TP={val_metrics.get('TP', 0)} FP={val_metrics.get('FP', 0)}")
    else:
        print("==> Phase 1: already completed, skipping to Phase 2")

    # Phase 2: unfreeze all, train
    for p in model.parameters():
        p.requires_grad = True
    opt = torch.optim.AdamW(model.parameters(), lr=lr_full, weight_decay=weight_decay)

    # Load optimizer state if resuming from Phase 2
    if resume and checkpoint is not None and resume_from_phase2:
        if 'optimizer_state' in checkpoint:
            opt.load_state_dict(checkpoint['optimizer_state'])
            print("Loaded optimizer state from checkpoint.")

    print("==> Phase 2: fine-tuning all layers")
    total_epochs = epochs_full

    # Calculate starting epoch for Phase 2
    if resume_from_phase2:
        phase2_start = (start_epoch - epochs_head) + 1
    else:
        phase2_start = 1

    for epoch in range(phase2_start, total_epochs+1):
        tr_loss = train_one_epoch(epoch, model, train_loader, opt, crit, device)
        val_metrics = evaluate_frame_level(model, val_loader, device, alpha=alpha)
        # Use weighted precision for model selection (better for class imbalance)
        score = val_metrics["wP"]
        is_best = score > best_val
        if is_best: best_val = score
        save_checkpoint({
            "epoch": epochs_head + epoch,
            "model_state": model.state_dict(),
            "optimizer_state": opt.state_dict(),
            "val_metrics": val_metrics,
            "args": {
                "data_dir": data_dir, "out_dir": out_dir, "epochs_head": epochs_head,
                "epochs_full": epochs_full, "batch_size": batch_size, "img_size": img_size,
                "num_workers": num_workers, "seed": seed, "lr_head": lr_head,
                "lr_full": lr_full, "weight_decay": weight_decay, "alpha": alpha
            },
        }, is_best, out_dir, name="xception")
        print(f"[FT {epoch}/{total_epochs}] loss={tr_loss:.4f}  wP={score:.4f}  acc={val_metrics['acc']:.4f}  prec={val_metrics.get('precision', 0):.4f}  rec={val_metrics.get('recall', 0):.4f}  log_loss={val_metrics.get('log_loss', float('nan')):.4f}  TP={val_metrics.get('TP', 0)} FP={val_metrics.get('FP', 0)}")

    print(f"Training complete. Best weighted precision (wP): {best_val:.4f}")

# Example usage (you can call this function in a separate cell):
data_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops' # Replace with your crops directory
out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc' # Replace with your desired output directory for runs

# Calculate alpha from your dataset if needed:
# For example, if your dataset has 1000 real:1000 fake (1:1 ratio) and organic traffic is 1M:1,
# then alpha = (1,000,000/1) / (1000/1000) = 1,000,000
# If unsure, start with alpha=1.0 (standard precision) or alpha=100-1000 for moderate imbalance

In [14]:
train_model(
    data_dir, out_dir,
    epochs_head=3,      # Full training
    epochs_full=15,     # Full training
    batch_size=32,
    alpha=1.0,
    use_class_weights=True  # Keep this enabled!
)

# For class imbalance, use higher alpha, e.g., alpha=1000 or alpha=10000

Using device: cuda
Using weighted precision with alpha=1.0 for model selection
  (alpha = ratio of real:fake in organic traffic / ratio of real:fake in dataset)
  (higher alpha penalizes false positives more, important for class imbalance)
Classes: ['fake_videos', 'real_videos'] (expect ['fake', 'real'] or similar two classes)
Class weights: [0.6210911 2.5645611] (higher weight = more penalty for misclassification)
  This will penalize false positives (real->fake) more during training
No checkpoint found. Starting training from scratch with pretrained ImageNet weights.


/usr/local/lib/python3.12/dist-packages/timm/models/_factory.py:138: UserWarning: Mapping deprecated model name xception to current legacy_xception.
  model = create_fn(


==> Phase 1: training the head
[Head 1/3] loss=0.6294  wP=0.4561  acc=0.7623  prec=0.4561  rec=0.5625  log_loss=0.5238  TP=135 FP=161
[Head 2/3] loss=0.5630  wP=0.3615  acc=0.6506  prec=0.3615  rec=0.8208  log_loss=0.6132  TP=197 FP=348
[Head 3/3] loss=0.5310  wP=0.4151  acc=0.7149  prec=0.4151  rec=0.8042  log_loss=0.5573  TP=193 FP=272
==> Phase 2: fine-tuning all layers
[FT 1/15] loss=0.1898  wP=0.8627  acc=0.9508  prec=0.8627  rec=0.9167  log_loss=0.1154  TP=220 FP=35
[FT 2/15] loss=0.0663  wP=0.8770  acc=0.9500  prec=0.8770  rec=0.8917  log_loss=0.1394  TP=214 FP=30
[FT 3/15] loss=0.0499  wP=0.7662  acc=0.9178  prec=0.7662  rec=0.8875  log_loss=0.2568  TP=213 FP=65
[FT 4/15] loss=0.0431  wP=0.9454  acc=0.9312  prec=0.9454  rec=0.7208  log_loss=0.2060  TP=173 FP=10
[FT 5/15] loss=0.0323  wP=0.9685  acc=0.9714  prec=0.9685  rec=0.8958  log_loss=0.0838  TP=215 FP=7
[FT 6/15] loss=0.0391  wP=0.8762  acc=0.9214  prec=0.8762  rec=0.7375  log_loss=0.2513  TP=177 FP=25
[FT 7/15] loss=0.03

In [39]:
import torch
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import os

# Define the path to the best model checkpoint
# Update this to match where you saved your model with class weights
out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc' # Model trained with class weights
# out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc' # Original model (commented out)
best_model_path = os.path.join(out_dir, 'xception_best_dec2.pt')

# Define the path to the test data
data_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops' # Make sure this matches your crops directory
test_data_dir = os.path.join(data_dir, 'test')

# Check if the best model checkpoint exists
if not os.path.exists(best_model_path):
    print(f"Error: Best model checkpoint not found at {best_model_path}")
else:
    # Load the best model state
    checkpoint = torch.load(best_model_path)

    # Build the model architecture (needs to match the one used for training)
    import timm
    model = timm.create_model("xception", pretrained=False, num_classes=2) # pretrained=False because we are loading a trained model
    model.load_state_dict(checkpoint['model_state'])

    # Set the model to evaluation mode
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = model.to(device)
    model.eval()
    print(f"Loaded best model from {best_model_path} and set to evaluation mode on {device}.")

    # Define the transformations for the test set (same as validation)
    img_size = 299
    test_tf = transforms.Compose([
        transforms.Resize(img_size),
        transforms.CenterCrop(img_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
    ])

    # Create the test dataset and dataloader
    try:
        test_ds = datasets.ImageFolder(test_data_dir, transform=test_tf)
        test_loader = DataLoader(test_ds, batch_size=32, shuffle=False, num_workers=8, pin_memory=True) # Use a reasonable batch size
        print(f"Created test dataset with {len(test_ds)} images.")

        # Get alpha from checkpoint args if available, otherwise use default
        alpha = checkpoint.get('args', {}).get('alpha', 1.0)
        print(f"Using alpha={alpha} for weighted precision (from checkpoint args or default)")

        # Using threshold 0.85 (optimal for class-weights model: 31 FP, 87.92% recall, 87.19% precision)
        test_threshold = 0.85
        print(f"Using threshold={test_threshold} for classification (optimized for FP reduction with class-weights model)")

        # Evaluate the model on the test set
        print("Evaluating model on the test set...")
        test_metrics = evaluate_frame_level(model, test_loader, device, alpha=alpha, threshold=test_threshold)

        print("\nTest Set Metrics (same format as training):")
        print(f"  Weighted Precision (wP): {test_metrics['wP']:.4f}")
        print(f"  Accuracy: {test_metrics['acc']:.4f}")
        print(f"  Precision: {test_metrics.get('precision', 0):.4f}")
        print(f"  Recall: {test_metrics.get('recall', 0):.4f}")
        print(f"  True Positives (TP): {test_metrics.get('TP', 0)}")
        print(f"  False Positives (FP): {test_metrics.get('FP', 0)}")
        print(f"  False Negatives (FN): {test_metrics.get('FN', 0)}")
        print(f"  True Negatives (TN): {test_metrics.get('TN', 0)}")
        print(f"\nAdditional Metrics:")
        print(f"  Log Loss: {test_metrics.get('log_loss', float('nan')):.4f}  (for DFDC comparison)")
        print(f"  F1 Score: {test_metrics['f1']:.4f}")
        print(f"  AUC: {test_metrics['auc']:.4f}")
        print(f"  AP (Fake): {test_metrics['ap_fake']:.4f}")

    except FileNotFoundError:
        print(f"Error: Test data directory not found at {test_data_dir}. Please ensure the path is correct.")
    except Exception as e:
        print(f"An error occurred during test set evaluation: {e}")

/usr/local/lib/python3.12/dist-packages/timm/models/_factory.py:138: UserWarning: Mapping deprecated model name xception to current legacy_xception.
  model = create_fn(


Loaded best model from /content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc/xception_best_dec2.pt and set to evaluation mode on cuda.
Created test dataset with 1159 images.
Using alpha=1.0 for weighted precision (from checkpoint args or default)
Using threshold=0.85 for classification (optimized for FP reduction with class-weights model)
Evaluating model on the test set...

Test Set Metrics (same format as training):
  Weighted Precision (wP): 0.8719
  Accuracy: 0.9482
  Precision: 0.8719
  Recall: 0.8792
  True Positives (TP): 211
  False Positives (FP): 31
  False Negatives (FN): 29
  True Negatives (TN): 888

Additional Metrics:
  Log Loss: 0.2304  (for DFDC comparison)
  F1 Score: 0.8755
  AUC: 0.9825
  AP (Fake): 0.9468


Analysis
Best for minimizing false positives: 0.8
Lowest FP: 30 (43% reduction)
Highest precision: 0.8770
Trade-off: recall drops to 89.17%
Best balance: 0.75
Good FP reduction: 34 (36% reduction)
Good precision: 0.8635
Better recall: 89.58% (vs 89.17% at 0.8)
Slightly better F1: 0.8793
Best recall: 0.7
Good FP reduction: 37 (30% reduction)
Highest recall: 90.42%
Good precision: 0.8543
Recommendation
Choose threshold 0.75 as a balanced option:
36% FP reduction (53 → 34)
Precision 0.8635
Recall 89.58%
Best F1 score

In [13]:
# Quick threshold tuning - test different thresholds without retraining
# This helps find the optimal threshold to reduce false positives

import torch
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import os
import numpy as np
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score

# Load your best model
out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc'
best_model_path = os.path.join(out_dir, 'xception_best.pt')
data_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops'
test_data_dir = os.path.join(data_dir, 'test')

if os.path.exists(best_model_path):
    checkpoint = torch.load(best_model_path)
    import timm
    model = timm.create_model("xception", pretrained=False, num_classes=2)
    model.load_state_dict(checkpoint['model_state'])
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = model.to(device)
    model.eval()

    # Prepare test data
    img_size = 299
    test_tf = transforms.Compose([
        transforms.Resize(img_size),
        transforms.CenterCrop(img_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
    ])
    test_ds = datasets.ImageFolder(test_data_dir, transform=test_tf)
    test_loader = DataLoader(test_ds, batch_size=32, shuffle=False, num_workers=8, pin_memory=True)

    # Get all predictions and probabilities
    all_probs = []
    all_targets = []
    with torch.no_grad():
        for x, y in test_loader:
            x = x.to(device)
            logits = model(x)
            probs = torch.softmax(logits, dim=1)[:, 1].cpu().numpy()
            all_probs.extend(probs)
            all_targets.extend(y.numpy())

    all_probs = np.array(all_probs)
    all_targets = np.array(all_targets)

    # Test different thresholds
    print("Threshold Tuning Results:")
    print("=" * 80)
    print(f"{'Threshold':<12} {'Precision':<12} {'Recall':<12} {'F1':<12} {'TP':<8} {'FP':<8} {'FN':<8} {'wP@alpha=1':<12}")
    print("=" * 80)

    thresholds = [0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
    alpha = 1.0  # You can change this to test with different alpha values

    for threshold in thresholds:
        preds = (all_probs >= threshold).astype(int)
        cm = confusion_matrix(all_targets, preds)
        if cm.shape == (2, 2):
            TN, FP, FN, TP = cm.ravel()
        else:
            TN, FP, FN, TP = 0, 0, 0, 0

        precision = precision_score(all_targets, preds) if (TP + FP) > 0 else 0.0
        recall = recall_score(all_targets, preds) if (TP + FN) > 0 else 0.0
        f1 = f1_score(all_targets, preds)

        # Calculate weighted precision
        if TP + alpha * FP > 0:
            wP = TP / (TP + alpha * FP)
        else:
            wP = 0.0

        print(f"{threshold:<12.2f} {precision:<12.4f} {recall:<12.4f} {f1:<12.4f} {TP:<8} {FP:<8} {FN:<8} {wP:<12.4f}")

    print("=" * 80)
    print("\nRecommendation: Choose threshold that balances low FP with acceptable recall")
    print("For deployment with class imbalance, prioritize lower FP (higher threshold)")
else:
    print(f"Best model not found at {best_model_path}")


/usr/local/lib/python3.12/dist-packages/timm/models/_factory.py:138: UserWarning: Mapping deprecated model name xception to current legacy_xception.
  model = create_fn(


Threshold Tuning Results:
Threshold    Precision    Recall       F1           TP       FP       FN       wP@alpha=1  
0.30         0.7692       0.9583       0.8534       230      69       10       0.7692      
0.40         0.7828       0.9458       0.8566       227      63       13       0.7828      
0.50         0.8094       0.9375       0.8687       225      53       15       0.8094      
0.60         0.8209       0.9167       0.8661       220      48       20       0.8209      
0.70         0.8543       0.9042       0.8785       217      37       23       0.8543      
0.80         0.8770       0.8917       0.8843       214      30       26       0.8770      
0.90         0.9052       0.8750       0.8898       210      22       30       0.9052      

Recommendation: Choose threshold that balances low FP with acceptable recall
For deployment with class imbalance, prioritize lower FP (higher threshold)


In [19]:
# Test different thresholds on the class-weights model to see if we can reduce FP
# This will help us find the best threshold for the test set

import torch
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import os
import numpy as np
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score

# Load the class-weights model
out_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/runs/xception_dfdc_classweights_test'
best_model_path = os.path.join(out_dir, 'xception_best_dec2.pt')
data_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops'
test_data_dir = os.path.join(data_dir, 'test')

if os.path.exists(best_model_path):
    checkpoint = torch.load(best_model_path)
    import timm
    model = timm.create_model("xception", pretrained=False, num_classes=2)
    model.load_state_dict(checkpoint['model_state'])
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = model.to(device)
    model.eval()

    # Prepare test data
    img_size = 299
    test_tf = transforms.Compose([
        transforms.Resize(img_size),
        transforms.CenterCrop(img_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
    ])
    test_ds = datasets.ImageFolder(test_data_dir, transform=test_tf)
    test_loader = DataLoader(test_ds, batch_size=32, shuffle=False, num_workers=8, pin_memory=True)

    # Get all predictions and probabilities
    all_probs = []
    all_targets = []
    with torch.no_grad():
        for x, y in test_loader:
            x = x.to(device)
            logits = model(x)
            probs = torch.softmax(logits, dim=1)[:, 1].cpu().numpy()
            all_probs.extend(probs)
            all_targets.extend(y.numpy())

    all_probs = np.array(all_probs)
    all_targets = np.array(all_targets)

    # Test different thresholds
    print("Threshold Tuning on TEST SET (Class-Weights Model):")
    print("=" * 80)
    print(f"{'Threshold':<12} {'Precision':<12} {'Recall':<12} {'F1':<12} {'TP':<8} {'FP':<8} {'FN':<8} {'wP@alpha=1':<12}")
    print("=" * 80)

    thresholds = [0.5, 0.6, 0.7, 0.75, 0.8, 0.85, 0.9]
    alpha = 1.0

    for threshold in thresholds:
        preds = (all_probs >= threshold).astype(int)
        cm = confusion_matrix(all_targets, preds)
        if cm.shape == (2, 2):
            TN, FP, FN, TP = cm.ravel()
        else:
            TN, FP, FN, TP = 0, 0, 0, 0

        precision = precision_score(all_targets, preds) if (TP + FP) > 0 else 0.0
        recall = recall_score(all_targets, preds) if (TP + FN) > 0 else 0.0
        f1 = f1_score(all_targets, preds)

        # Calculate weighted precision
        if TP + alpha * FP > 0:
            wP = TP / (TP + alpha * FP)
        else:
            wP = 0.0

        print(f"{threshold:<12.2f} {precision:<12.4f} {recall:<12.4f} {f1:<12.4f} {TP:<8} {FP:<8} {FN:<8} {wP:<12.4f}")

    print("=" * 80)
    print("\nThis shows how the class-weights model performs on TEST set with different thresholds")
    print("Compare with original model results to see if class weights helped")
else:
    print(f"Model not found at {best_model_path}")

/usr/local/lib/python3.12/dist-packages/timm/models/_factory.py:138: UserWarning: Mapping deprecated model name xception to current legacy_xception.
  model = create_fn(


Threshold Tuning on TEST SET (Class-Weights Model):
Threshold    Precision    Recall       F1           TP       FP       FN       wP@alpha=1  
0.50         0.8151       0.9000       0.8554       216      49       24       0.8151      
0.60         0.8168       0.8917       0.8526       214      48       26       0.8168      
0.70         0.8346       0.8833       0.8583       212      42       28       0.8346      
0.75         0.8514       0.8833       0.8671       212      37       28       0.8514      
0.80         0.8583       0.8833       0.8706       212      35       28       0.8583      
0.85         0.8719       0.8792       0.8755       211      31       29       0.8719      
0.90         0.8851       0.8667       0.8758       208      27       32       0.8851      

This shows how the class-weights model performs on TEST set with different thresholds
Compare with original model results to see if class weights helped


In [45]:
# ============================================================================
# CRITICAL FIX: Class Index Bug
# ============================================================================
# The model uses: Class 0 = FAKE, Class 1 = REAL
# But we were using [:, 1] which is REAL, not FAKE!
# This explains why real videos had high "fake" probabilities
# ============================================================================

print("="*80)
print("FIXING CLASS INDEX BUG")
print("="*80)
print("Class mapping:")
print("  Class 0 = FAKE")
print("  Class 1 = REAL")
print("\nWe were using [:, 1] (REAL class) instead of [:, 0] (FAKE class)")
print("This is why predictions were backwards!")
print("="*80)

# Update the predict_video_with_ground_truth function
def predict_video_fixed(video_path, test_frames_dir='/content/drive/MyDrive/DeepfakeData/SampleData/crops/test', threshold=0.85):
    """
    FIXED VERSION: Uses correct class index (Class 0 = FAKE)
    """
    # Extract video name from path
    if os.path.isfile(video_path):
        video_name = Path(video_path).stem
    else:
        video_name = video_path

    print(f"Looking for frames from video: {video_name}")

    # Search for frames
    test_dir = Path(test_frames_dir)
    real_pattern = test_dir / 'real_videos' / f'{video_name}_*.jpg'
    fake_pattern = test_dir / 'fake_videos' / f'{video_name}_*.jpg'

    real_frames = glob.glob(str(real_pattern))
    fake_frames = glob.glob(str(fake_pattern))

    # Determine ground truth
    if real_frames and not fake_frames:
        frame_files = real_frames
        ground_truth = 'real'
    elif fake_frames and not real_frames:
        frame_files = fake_frames
        ground_truth = 'fake'
    elif real_frames and fake_frames:
        frame_files = real_frames if len(real_frames) >= len(fake_frames) else fake_frames
        ground_truth = 'real' if len(real_frames) >= len(fake_frames) else 'fake'
        print(f"Warning: Found frames in both folders. Using {ground_truth} folder ({len(frame_files)} frames)")
    else:
        return {
            'error': f'No frames found for video "{video_name}" in {test_frames_dir}',
            'prediction': None,
            'ground_truth': None
        }

    frame_files.sort()
    print(f"Found {len(frame_files)} frames (Ground truth: {ground_truth})")

    # Load and preprocess frames
    frames = []
    for frame_path in frame_files:
        img = Image.open(frame_path).convert('RGB')
        frames.append(val_tf(img))

    frames_tensor = torch.stack(frames).to(device)

    # Run predictions - FIXED: Use [:, 0] for FAKE class
    model.eval()
    with torch.no_grad():
        logits = model(frames_tensor)
        probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # FIXED: Class 0 = FAKE

    # Aggregate predictions
    avg_prob = float(np.mean(probs))
    fake_frames_count = np.sum(probs >= threshold)
    total_frames = len(probs)
    fake_ratio = fake_frames_count / total_frames

    # Majority vote: if >50% of frames are fake, video is fake
    prediction = 'fake' if fake_ratio > 0.5 else 'real'

    # Check if prediction is correct
    is_correct = (prediction == ground_truth)

    # Print results
    print(f"\n{'='*60}")
    print(f"RESULTS (FIXED):")
    print(f"{'='*60}")
    print(f"Ground Truth: {ground_truth.upper()}")
    print(f"Prediction:    {prediction.upper()}")
    print(f"Correct:       {'✓ YES' if is_correct else '✗ NO'}")
    print(f"{'='*60}")
    print(f"Average FAKE probability: {avg_prob:.4f}")
    print(f"Fake ratio: {fake_ratio:.4f} ({fake_frames_count}/{total_frames} frames)")
    print(f"Total frames analyzed: {total_frames}")
    print(f"{'='*60}\n")

    return {
        'prediction': prediction,
        'ground_truth': ground_truth,
        'is_correct': is_correct,
        'avg_prob': avg_prob,
        'fake_ratio': fake_ratio,
        'total_frames': total_frames,
        'confidence': abs(avg_prob - 0.5) * 2
    }

# Override the old function
predict_video_with_ground_truth = predict_video_fixed
predict_video = predict_video_fixed

print("\n✓ Fixed predict_video function!")
print("Now using Class 0 (FAKE) instead of Class 1 (REAL)")
print("\nTest it with: predict_video('video_name')")

FIXING CLASS INDEX BUG
Class mapping:
  Class 0 = FAKE
  Class 1 = REAL

We were using [:, 1] (REAL class) instead of [:, 0] (FAKE class)
This is why predictions were backwards!

✓ Fixed predict_video function!
Now using Class 0 (FAKE) instead of Class 1 (REAL)

Test it with: predict_video('video_name')


In [46]:


# ============================================================================
# Example Usage
# ============================================================================
# Example 1: Pass video name (frames should be in test folder)
result = predict_video_with_ground_truth('avvdgsennp')
print(f"Prediction: {result['prediction']}")
print(f"Average fake probability: {result['avg_prob']:.4f}")
print(f"Fake ratio: {result['fake_ratio']:.4f}")


Looking for frames from video: avvdgsennp
Found 6 frames (Ground truth: fake)

RESULTS (FIXED):
Ground Truth: FAKE
Prediction:    FAKE
Correct:       ✓ YES
Average FAKE probability: 0.9972
Fake ratio: 1.0000 (6/6 frames)
Total frames analyzed: 6

Prediction: fake
Average fake probability: 0.9972
Fake ratio: 1.0000


In [47]:
# ============================================================================
# TEST MODEL ON VALIDATION SET (FIXED VERSION)
# ============================================================================
# Using correct class index: Class 0 = FAKE, Class 1 = REAL

def get_all_val_video_names(val_frames_dir='/content/drive/MyDrive/DeepfakeData/SampleData/crops/val'):
    """
    Get all fake and real video names from validation set
    """
    val_dir = Path(val_frames_dir)

    # Get fake videos
    fake_dir = val_dir / 'fake_videos'
    fake_frame_files = list(fake_dir.glob('*.jpg')) if fake_dir.exists() else []
    fake_video_names = set()
    for frame_file in fake_frame_files:
        video_name = frame_file.stem.rsplit('_', 1)[0]
        fake_video_names.add(video_name)

    # Get real videos
    real_dir = val_dir / 'real_videos'
    real_frame_files = list(real_dir.glob('*.jpg')) if real_dir.exists() else []
    real_video_names = set()
    for frame_file in real_frame_files:
        video_name = frame_file.stem.rsplit('_', 1)[0]
        real_video_names.add(video_name)

    return sorted(list(fake_video_names)), sorted(list(real_video_names))

def test_videos_on_validation_set_fixed(fake_video_names, real_video_names, threshold=0.85):
    """
    Test model on validation set videos - FIXED VERSION (uses Class 0 = FAKE)
    """
    val_frames_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops/val'
    val_dir = Path(val_frames_dir)

    results = {'fake': [], 'real': []}

    # Test fake videos
    print("Testing FAKE videos in validation set...")
    for video_name in fake_video_names:
        fake_pattern = val_dir / 'fake_videos' / f'{video_name}_*.jpg'
        frame_files = sorted(glob.glob(str(fake_pattern)))

        if not frame_files:
            continue

        # Load and predict
        frames = []
        for frame_path in frame_files:
            img = Image.open(frame_path).convert('RGB')
            frames.append(val_tf(img))

        frames_tensor = torch.stack(frames).to(device)
        model.eval()
        with torch.no_grad():
            logits = model(frames_tensor)
            probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # FIXED: Class 0 = FAKE

        avg_prob = float(np.mean(probs))
        fake_frames = np.sum(probs >= threshold)
        fake_ratio = fake_frames / len(probs)
        prediction = 'fake' if fake_ratio > 0.5 else 'real'

        results['fake'].append({
            'video_name': video_name,
            'prediction': prediction,
            'ground_truth': 'fake',
            'is_correct': prediction == 'fake',
            'avg_prob': avg_prob,
            'fake_ratio': fake_ratio,
            'total_frames': len(probs)
        })

    # Test real videos
    print("Testing REAL videos in validation set...")
    for video_name in real_video_names:
        real_pattern = val_dir / 'real_videos' / f'{video_name}_*.jpg'
        frame_files = sorted(glob.glob(str(real_pattern)))

        if not frame_files:
            continue

        # Load and predict
        frames = []
        for frame_path in frame_files:
            img = Image.open(frame_path).convert('RGB')
            frames.append(val_tf(img))

        frames_tensor = torch.stack(frames).to(device)
        model.eval()
        with torch.no_grad():
            logits = model(frames_tensor)
            probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # FIXED: Class 0 = FAKE

        avg_prob = float(np.mean(probs))
        fake_frames = np.sum(probs >= threshold)
        fake_ratio = fake_frames / len(probs)
        prediction = 'fake' if fake_ratio > 0.5 else 'real'

        results['real'].append({
            'video_name': video_name,
            'prediction': prediction,
            'ground_truth': 'real',
            'is_correct': prediction == 'real',
            'avg_prob': avg_prob,
            'fake_ratio': fake_ratio,
            'total_frames': len(probs)
        })

    return results

# Get validation video names
print("="*80)
print("GETTING VALIDATION SET VIDEO NAMES")
print("="*80)

val_fake_names, val_real_names = get_all_val_video_names()

print(f"Found {len(val_fake_names)} fake videos in validation set")
print(f"Found {len(val_real_names)} real videos in validation set")

if len(val_fake_names) > 0:
    print(f"\nFirst 5 fake videos: {val_fake_names[:5]}")
if len(val_real_names) > 0:
    print(f"First 5 real videos: {val_real_names[:5]}")

# Test on validation set
print("\n" + "="*80)
print("TESTING MODEL ON VALIDATION SET (FIXED VERSION)")
print("="*80)

val_results = test_videos_on_validation_set_fixed(val_fake_names, val_real_names, threshold=0.85)

# Calculate metrics
fake_results = val_results['fake']
real_results = val_results['real']

fake_correct = sum(1 for r in fake_results if r['is_correct'])
fake_total = len(fake_results)
fake_accuracy = fake_correct / fake_total if fake_total > 0 else 0

real_correct = sum(1 for r in real_results if r['is_correct'])
real_total = len(real_results)
real_accuracy = real_correct / real_total if real_total > 0 else 0

total_correct = fake_correct + real_correct
total_videos = fake_total + real_total
overall_accuracy = total_correct / total_videos if total_videos > 0 else 0

print(f"\n{'='*80}")
print("VALIDATION SET RESULTS (FIXED - threshold=0.85)")
print(f"{'='*80}")
print(f"\nFAKE Videos:")
print(f"  Total: {fake_total}")
print(f"  Correctly identified as FAKE: {fake_correct}")
print(f"  Incorrectly identified as REAL: {fake_total - fake_correct}")
print(f"  Accuracy: {fake_accuracy:.4f} ({fake_accuracy*100:.2f}%)")

print(f"\nREAL Videos:")
print(f"  Total: {real_total}")
print(f"  Correctly identified as REAL: {real_correct}")
print(f"  Incorrectly identified as FAKE: {real_total - real_correct}")
print(f"  Accuracy: {real_accuracy:.4f} ({real_accuracy*100:.2f}%)")

print(f"\nOVERALL:")
print(f"  Total videos: {total_videos}")
print(f"  Correct predictions: {total_correct}")
print(f"  Overall accuracy: {overall_accuracy:.4f} ({overall_accuracy*100:.2f}%)")

# Show failed fake videos
failed_fake = [r for r in fake_results if not r['is_correct']]
if failed_fake:
    print(f"\n{'='*80}")
    print(f"FAILED FAKE VIDEOS IN VALIDATION ({len(failed_fake)}):")
    print(f"{'='*80}")
    for r in failed_fake[:10]:  # Show first 10
        print(f"  {r['video_name']}: Predicted {r['prediction']} (avg_prob={r['avg_prob']:.4f}, fake_ratio={r['fake_ratio']:.4f})")
    if len(failed_fake) > 10:
        print(f"  ... and {len(failed_fake) - 10} more")

# Show failed real videos (false positives)
failed_real = [r for r in real_results if not r['is_correct']]
if failed_real:
    print(f"\n{'='*80}")
    print(f"FALSE POSITIVES - REAL VIDEOS PREDICTED AS FAKE ({len(failed_real)}):")
    print(f"{'='*80}")
    for r in failed_real[:10]:  # Show first 10
        print(f"  {r['video_name']}: Predicted {r['prediction']} (avg_prob={r['avg_prob']:.4f}, fake_ratio={r['fake_ratio']:.4f})")
    if len(failed_real) > 10:
        print(f"  ... and {len(failed_real) - 10} more")

print(f"\n{'='*80}")
print("COMPARISON:")
print(f"{'='*80}")
print("Now with the FIXED class index, the model should perform much better!")
print("If validation accuracy is HIGH:")
print("  → Model is working correctly")
print("  → Test set may have different/harder fake videos")
print("If validation accuracy is still LOW:")
print("  → Model may need more training or different approach")
print("="*80)


GETTING VALIDATION SET VIDEO NAMES
Found 32 fake videos in validation set
Found 8 real videos in validation set

First 5 fake videos: ['abqwwspghj', 'acxwigylke', 'akvmwkdyuv', 'amaivqofda', 'aslsvlvpth']
First 5 real videos: ['abarnvbtwb', 'avmjormvsx', 'bejhvclboh', 'ccfoszqabv', 'cppdvdejkc']

TESTING MODEL ON VALIDATION SET (FIXED VERSION)
Testing FAKE videos in validation set...
Testing REAL videos in validation set...

VALIDATION SET RESULTS (FIXED - threshold=0.85)

FAKE Videos:
  Total: 32
  Correctly identified as FAKE: 32
  Incorrectly identified as REAL: 0
  Accuracy: 1.0000 (100.00%)

REAL Videos:
  Total: 8
  Correctly identified as REAL: 8
  Incorrectly identified as FAKE: 0
  Accuracy: 1.0000 (100.00%)

OVERALL:
  Total videos: 40
  Correct predictions: 40
  Overall accuracy: 1.0000 (100.00%)

COMPARISON:
Now with the FIXED class index, the model should perform much better!
If validation accuracy is HIGH:
  → Model is working correctly
  → Test set may have different/har

In [48]:
# ============================================================================
# TEST MODEL ON TEST SET (FIXED VERSION)
# ============================================================================
# Now let's test on the test set with the fixed class index

def test_videos_on_test_set_fixed(fake_video_names, real_video_names, threshold=0.85):
    """
    Test model on test set videos - FIXED VERSION (uses Class 0 = FAKE)
    """
    test_frames_dir = '/content/drive/MyDrive/DeepfakeData/SampleData/crops/test'
    test_dir = Path(test_frames_dir)

    results = {'fake': [], 'real': []}

    # Test fake videos
    print("Testing FAKE videos in test set...")
    for video_name in fake_video_names:
        fake_pattern = test_dir / 'fake_videos' / f'{video_name}_*.jpg'
        frame_files = sorted(glob.glob(str(fake_pattern)))

        if not frame_files:
            continue

        # Load and predict
        frames = []
        for frame_path in frame_files:
            img = Image.open(frame_path).convert('RGB')
            frames.append(val_tf(img))

        frames_tensor = torch.stack(frames).to(device)
        model.eval()
        with torch.no_grad():
            logits = model(frames_tensor)
            probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # FIXED: Class 0 = FAKE

        avg_prob = float(np.mean(probs))
        fake_frames = np.sum(probs >= threshold)
        fake_ratio = fake_frames / len(probs)
        prediction = 'fake' if fake_ratio > 0.5 else 'real'

        results['fake'].append({
            'video_name': video_name,
            'prediction': prediction,
            'ground_truth': 'fake',
            'is_correct': prediction == 'fake',
            'avg_prob': avg_prob,
            'fake_ratio': fake_ratio,
            'total_frames': len(probs)
        })

    # Test real videos
    print("Testing REAL videos in test set...")
    for video_name in real_video_names:
        real_pattern = test_dir / 'real_videos' / f'{video_name}_*.jpg'
        frame_files = sorted(glob.glob(str(real_pattern)))

        if not frame_files:
            continue

        # Load and predict
        frames = []
        for frame_path in frame_files:
            img = Image.open(frame_path).convert('RGB')
            frames.append(val_tf(img))

        frames_tensor = torch.stack(frames).to(device)
        model.eval()
        with torch.no_grad():
            logits = model(frames_tensor)
            probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # FIXED: Class 0 = FAKE

        avg_prob = float(np.mean(probs))
        fake_frames = np.sum(probs >= threshold)
        fake_ratio = fake_frames / len(probs)
        prediction = 'fake' if fake_ratio > 0.5 else 'real'

        results['real'].append({
            'video_name': video_name,
            'prediction': prediction,
            'ground_truth': 'real',
            'is_correct': prediction == 'real',
            'avg_prob': avg_prob,
            'fake_ratio': fake_ratio,
            'total_frames': len(probs)
        })

    return results

# Get test video names (using the function from earlier)
print("="*80)
print("GETTING TEST SET VIDEO NAMES")
print("="*80)

test_fake_names = get_all_fake_video_names('/content/drive/MyDrive/DeepfakeData/SampleData/crops/test')

# Get real video names from test set
test_dir = Path('/content/drive/MyDrive/DeepfakeData/SampleData/crops/test')
real_dir = test_dir / 'real_videos'
real_frame_files = list(real_dir.glob('*.jpg')) if real_dir.exists() else []
test_real_names = set()
for frame_file in real_frame_files:
    video_name = frame_file.stem.rsplit('_', 1)[0]
    test_real_names.add(video_name)
test_real_names = sorted(list(test_real_names))

print(f"Found {len(test_fake_names)} fake videos in test set")
print(f"Found {len(test_real_names)} real videos in test set")

if len(test_fake_names) > 0:
    print(f"\nFirst 5 fake videos: {test_fake_names[:5]}")
if len(test_real_names) > 0:
    print(f"First 5 real videos: {test_real_names[:5]}")

# Test on test set
print("\n" + "="*80)
print("TESTING MODEL ON TEST SET (FIXED VERSION)")
print("="*80)

test_results = test_videos_on_test_set_fixed(test_fake_names, test_real_names, threshold=0.85)

# Calculate metrics
fake_results = test_results['fake']
real_results = test_results['real']

fake_correct = sum(1 for r in fake_results if r['is_correct'])
fake_total = len(fake_results)
fake_accuracy = fake_correct / fake_total if fake_total > 0 else 0

real_correct = sum(1 for r in real_results if r['is_correct'])
real_total = len(real_results)
real_accuracy = real_correct / real_total if real_total > 0 else 0

total_correct = fake_correct + real_correct
total_videos = fake_total + real_total
overall_accuracy = total_correct / total_videos if total_videos > 0 else 0

print(f"\n{'='*80}")
print("TEST SET RESULTS (FIXED - threshold=0.85)")
print(f"{'='*80}")
print(f"\nFAKE Videos:")
print(f"  Total: {fake_total}")
print(f"  Correctly identified as FAKE: {fake_correct}")
print(f"  Incorrectly identified as REAL: {fake_total - fake_correct}")
print(f"  Accuracy: {fake_accuracy:.4f} ({fake_accuracy*100:.2f}%)")

print(f"\nREAL Videos:")
print(f"  Total: {real_total}")
print(f"  Correctly identified as REAL: {real_correct}")
print(f"  Incorrectly identified as FAKE: {real_total - real_correct}")
print(f"  Accuracy: {real_accuracy:.4f} ({real_accuracy*100:.2f}%)")

print(f"\nOVERALL:")
print(f"  Total videos: {total_videos}")
print(f"  Correct predictions: {total_correct}")
print(f"  Overall accuracy: {overall_accuracy:.4f} ({overall_accuracy*100:.2f}%)")

# Show failed fake videos
failed_fake = [r for r in fake_results if not r['is_correct']]
if failed_fake:
    print(f"\n{'='*80}")
    print(f"FAILED FAKE VIDEOS IN TEST ({len(failed_fake)}):")
    print(f"{'='*80}")
    for r in failed_fake:
        print(f"  {r['video_name']}: Predicted {r['prediction']} (avg_prob={r['avg_prob']:.4f}, fake_ratio={r['fake_ratio']:.4f})")

# Show failed real videos (false positives)
failed_real = [r for r in real_results if not r['is_correct']]
if failed_real:
    print(f"\n{'='*80}")
    print(f"FALSE POSITIVES - REAL VIDEOS PREDICTED AS FAKE ({len(failed_real)}):")
    print(f"{'='*80}")
    for r in failed_real:
        print(f"  {r['video_name']}: Predicted {r['prediction']} (avg_prob={r['avg_prob']:.4f}, fake_ratio={r['fake_ratio']:.4f})")

print(f"\n{'='*80}")
print("SUMMARY:")
print(f"{'='*80}")
print(f"Validation Set: 100.00% accuracy (40/40 videos)")
print(f"Test Set: {overall_accuracy*100:.2f}% accuracy ({total_correct}/{total_videos} videos)")
print(f"\nIf test accuracy is close to validation:")
print("  → Model generalizes well!")
print("If test accuracy is lower:")
print("  → Test set may have different/harder fake videos")
print("="*80)


GETTING TEST SET VIDEO NAMES
Found 33 fake videos in test folder
Found 33 fake videos in test set
Found 8 real videos in test set

First 5 fake videos: ['acqfdwsrhi', 'adhsbajydo', 'agqphdxmwt', 'arkroixhey', 'asdpeebotb']
First 5 real videos: ['ahqqqilsxt', 'aybgughjxh', 'bwhlgysghg', 'chtapglbcj', 'crezycjqyk']

TESTING MODEL ON TEST SET (FIXED VERSION)
Testing FAKE videos in test set...
Testing REAL videos in test set...

TEST SET RESULTS (FIXED - threshold=0.85)

FAKE Videos:
  Total: 33
  Correctly identified as FAKE: 32
  Incorrectly identified as REAL: 1
  Accuracy: 0.9697 (96.97%)

REAL Videos:
  Total: 8
  Correctly identified as REAL: 7
  Incorrectly identified as FAKE: 1
  Accuracy: 0.8750 (87.50%)

OVERALL:
  Total videos: 41
  Correct predictions: 39
  Overall accuracy: 0.9512 (95.12%)

FAILED FAKE VIDEOS IN TEST (1):
  bqnymlsayl: Predicted real (avg_prob=0.7171, fake_ratio=0.5000)

FALSE POSITIVES - REAL VIDEOS PREDICTED AS FAKE (1):
  bwhlgysghg: Predicted fake (avg_pro

In [53]:
# ============================================================================
# VIDEO-LEVEL PREDICTION: Combine Frames by Video Name
# ============================================================================
# Groups frames by video name and aggregates predictions to determine if video is fake or real

def predict_video_from_frames(video_name, frames_dir='/content/drive/MyDrive/DeepfakeData/CompleteData/organized/crops/test', threshold=0.85):
    """
    Predict if a video is fake or real by combining all frames from that video.

    Args:
        video_name: Name of the video (e.g., 'ahqqqilsxt')
        frames_dir: Base directory containing frames (looks in both real_videos and fake_videos)
        threshold: Probability threshold for frame classification

    Returns:
        Dictionary with video-level prediction and frame statistics
    """
    frames_base = Path(frames_dir)

    # Search for frames in both real and fake folders
    real_pattern = frames_base / 'real_videos' / f'{video_name}_*.jpg'
    fake_pattern = frames_base / 'fake_videos' / f'{video_name}_*.jpg'

    real_frames = glob.glob(str(real_pattern))
    fake_frames = glob.glob(str(fake_pattern))

    # Determine ground truth and get frame files
    if real_frames and not fake_frames:
        frame_files = real_frames
        ground_truth = 'real'
    elif fake_frames and not real_frames:
        frame_files = fake_frames
        ground_truth = 'fake'
    elif real_frames and fake_frames:
        frame_files = real_frames if len(real_frames) >= len(fake_frames) else fake_frames
        ground_truth = 'real' if len(real_frames) >= len(fake_frames) else 'fake'
    else:
        return {
            'error': f'No frames found for video "{video_name}"',
            'video_name': video_name,
            'prediction': None
        }

    frame_files.sort()

    # Load and preprocess all frames
    frames = []
    for frame_path in frame_files:
        img = Image.open(frame_path).convert('RGB')
        frames.append(val_tf(img))

    frames_tensor = torch.stack(frames).to(device)

    # Get frame-level predictions
    model.eval()
    with torch.no_grad():
        logits = model(frames_tensor)
        probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # Class 0 = FAKE

    # Frame-level statistics
    total_frames = len(probs)
    avg_fake_prob = float(np.mean(probs))
    max_fake_prob = float(np.max(probs))
    min_fake_prob = float(np.min(probs))

    # Count frames above threshold
    frames_above_threshold = np.sum(probs >= threshold)
    fake_ratio = frames_above_threshold / total_frames

    # Video-level prediction: Majority vote
    # If >50% of frames are predicted as fake, video is fake
    video_prediction = 'fake' if fake_ratio > 0.5 else 'real'

    # Check if correct
    is_correct = (video_prediction == ground_truth)

    return {
        'video_name': video_name,
        'ground_truth': ground_truth,
        'prediction': video_prediction,
        'is_correct': is_correct,
        'total_frames': total_frames,
        'avg_fake_prob': avg_fake_prob,
        'max_fake_prob': max_fake_prob,
        'min_fake_prob': min_fake_prob,
        'frames_above_threshold': int(frames_above_threshold),
        'fake_ratio': fake_ratio,
        'frame_probabilities': probs.tolist()
    }

def analyze_all_videos_in_directory(frames_dir='/content/drive/MyDrive/DeepfakeData/CompleteData/organized/crops/test', threshold=0.85):
    """
    Analyze all videos in a directory by combining frames by video name.

    Args:
        frames_dir: Directory containing frames (with real_videos and fake_videos subdirs)
        threshold: Probability threshold for frame classification

    Returns:
        Dictionary with results for all videos
    """
    frames_base = Path(frames_dir)

    # Get all unique video names from both folders
    all_video_names = set()

    # From fake_videos folder
    fake_dir = frames_base / 'fake_videos'
    if fake_dir.exists():
        for frame_file in fake_dir.glob('*.jpg'):
            video_name = frame_file.stem.rsplit('_', 1)[0]
            all_video_names.add(video_name)

    # From real_videos folder
    real_dir = frames_base / 'real_videos'
    if real_dir.exists():
        for frame_file in real_dir.glob('*.jpg'):
            video_name = frame_file.stem.rsplit('_', 1)[0]
            all_video_names.add(video_name)

    all_video_names = sorted(list(all_video_names))

    print(f"Found {len(all_video_names)} unique videos")
    print(f"Analyzing videos with threshold={threshold}...")
    print("="*80)

    # Analyze each video
    results = []
    for video_name in all_video_names:
        result = predict_video_from_frames(video_name, frames_dir, threshold)
        if 'error' not in result:
            results.append(result)

    return results

# ============================================================================
# Analyze Test Set Videos
# ============================================================================
print("="*80)
print("VIDEO-LEVEL ANALYSIS: Combining Frames by Video Name")
print("="*80)
print("This will group all frames by video name and determine if each video is fake or real")
print("="*80)

# Analyze all videos in test set
test_results = analyze_all_videos_in_directory(
    frames_dir='/content/drive/MyDrive/DeepfakeData/CompleteData/organized/crops/test',
    threshold=0.85
)

# Separate by ground truth
fake_videos = [r for r in test_results if r['ground_truth'] == 'fake']
real_videos = [r for r in test_results if r['ground_truth'] == 'real']

# Calculate metrics
fake_correct = sum(1 for r in fake_videos if r['is_correct'])
fake_total = len(fake_videos)
fake_accuracy = fake_correct / fake_total if fake_total > 0 else 0

real_correct = sum(1 for r in real_videos if r['is_correct'])
real_total = len(real_videos)
real_accuracy = real_correct / real_total if real_total > 0 else 0

total_correct = fake_correct + real_correct
total_videos = len(test_results)
overall_accuracy = total_correct / total_videos if total_videos > 0 else 0

# Print summary
print(f"\n{'='*80}")
print("VIDEO-LEVEL RESULTS (Frames Combined by Video Name)")
print(f"{'='*80}")
print(f"\nTotal Videos Analyzed: {total_videos}")
print(f"  Fake videos: {fake_total}")
print(f"  Real videos: {real_total}")

print(f"\nFAKE Videos:")
print(f"  Correctly identified as FAKE: {fake_correct}/{fake_total}")
print(f"  Accuracy: {fake_accuracy:.4f} ({fake_accuracy*100:.2f}%)")

print(f"\nREAL Videos:")
print(f"  Correctly identified as REAL: {real_correct}/{real_total}")
print(f"  Accuracy: {real_accuracy:.4f} ({real_accuracy*100:.2f}%)")

print(f"\nOVERALL:")
print(f"  Correct predictions: {total_correct}/{total_videos}")
print(f"  Overall accuracy: {overall_accuracy:.4f} ({overall_accuracy*100:.2f}%)")

# Show detailed results for each video
print(f"\n{'='*80}")
print("DETAILED RESULTS FOR EACH VIDEO:")
print(f"{'='*80}")
print(f"{'Video Name':<20} {'GT':<6} {'Pred':<6} {'Correct':<8} {'Avg Prob':<10} {'Fake Ratio':<12} {'Frames':<8}")
print("-"*80)

for result in sorted(test_results, key=lambda x: (x['ground_truth'], x['video_name'])):
    correct_str = "✓ YES" if result['is_correct'] else "✗ NO"
    print(f"{result['video_name']:<20} {result['ground_truth']:<6} {result['prediction']:<6} {correct_str:<8} {result['avg_fake_prob']:<10.4f} {result['fake_ratio']:<12.4f} {result['total_frames']:<8}")

# Show failed videos
failed_videos = [r for r in test_results if not r['is_correct']]
if failed_videos:
    print(f"\n{'='*80}")
    print(f"FAILED VIDEOS ({len(failed_videos)}):")
    print(f"{'='*80}")
    for r in failed_videos:
        print(f"  {r['video_name']}: Ground truth={r['ground_truth']}, Predicted={r['prediction']}")
        print(f"    Avg fake prob: {r['avg_fake_prob']:.4f}, Fake ratio: {r['fake_ratio']:.4f}, Frames: {r['total_frames']}")

print("\n" + "="*80)
print("Analysis complete!")
print("="*80)


VIDEO-LEVEL ANALYSIS: Combining Frames by Video Name
This will group all frames by video name and determine if each video is fake or real
Found 134 unique videos
Analyzing videos with threshold=0.85...

VIDEO-LEVEL RESULTS (Frames Combined by Video Name)

Total Videos Analyzed: 134
  Fake videos: 125
  Real videos: 9

FAKE Videos:
  Correctly identified as FAKE: 125/125
  Accuracy: 1.0000 (100.00%)

REAL Videos:
  Correctly identified as REAL: 0/9
  Accuracy: 0.0000 (0.00%)

OVERALL:
  Correct predictions: 125/134
  Overall accuracy: 0.9328 (93.28%)

DETAILED RESULTS FOR EACH VIDEO:
Video Name           GT     Pred   Correct  Avg Prob   Fake Ratio   Frames  
--------------------------------------------------------------------------------
acdkfksyev           fake   fake   ✓ YES    0.9955     1.0000       28      
achdeirhym           fake   fake   ✓ YES    0.9867     0.9667       30      
aiuimrebtg           fake   fake   ✓ YES    0.9605     0.9667       30      
bkcztuqxbp           

In [52]:
# ============================================================================
# DFDC WINNER'S CONFIDENT STRATEGY
# ============================================================================
# Implementation of the aggregation strategy from DFDC challenge winner
# This uses high-confidence frames instead of simple averaging

import numpy as np

def confident_strategy(pred, t=0.8):
    """
    DFDC winner's confident aggregation strategy.

    Args:
        pred: Array of frame-level fake probabilities
        t: Threshold for high-confidence fake predictions (default 0.8)

    Returns:
        Aggregated fake probability for the video
    """
    pred = np.array(pred)
    sz = len(pred)
    fakes = np.count_nonzero(pred > t)

    # If many frames are detected as fakes with high probability
    if fakes > sz // 2.5 and fakes > 11:
        return np.mean(pred[pred > t])
    # If most frames are clearly real (low fake probability)
    elif np.count_nonzero(pred < 0.2) > 0.9 * sz:
        return np.mean(pred[pred < 0.2])
    # Otherwise, use mean of all frames
    else:
        return np.mean(pred)

def predict_video_with_confident_strategy(video_name, frames_dir='/content/drive/MyDrive/DeepfakeData/CompleteData/organized/crops/test', threshold=0.85, confident_t=0.8):
    """
    Predict video using DFDC winner's confident strategy.

    Args:
        video_name: Name of the video
        frames_dir: Directory containing frames
        threshold: Final classification threshold (for fake/real decision)
        confident_t: Threshold for high-confidence frames in aggregation (default 0.8)

    Returns:
        Dictionary with prediction and statistics
    """
    frames_base = Path(frames_dir)

    # Search for frames
    real_pattern = frames_base / 'real_videos' / f'{video_name}_*.jpg'
    fake_pattern = frames_base / 'fake_videos' / f'{video_name}_*.jpg'

    real_frames = glob.glob(str(real_pattern))
    fake_frames = glob.glob(str(fake_pattern))

    # Determine ground truth and get frame files
    if real_frames and not fake_frames:
        frame_files = real_frames
        ground_truth = 'real'
    elif fake_frames and not real_frames:
        frame_files = fake_frames
        ground_truth = 'fake'
    elif real_frames and fake_frames:
        frame_files = real_frames if len(real_frames) >= len(fake_frames) else fake_frames
        ground_truth = 'real' if len(real_frames) >= len(fake_frames) else 'fake'
    else:
        return {
            'error': f'No frames found for video "{video_name}"',
            'video_name': video_name,
            'prediction': None
        }

    frame_files.sort()

    # Load and preprocess all frames
    frames = []
    for frame_path in frame_files:
        img = Image.open(frame_path).convert('RGB')
        frames.append(val_tf(img))

    frames_tensor = torch.stack(frames).to(device)

    # Get frame-level predictions
    model.eval()
    with torch.no_grad():
        logits = model(frames_tensor)
        probs = torch.softmax(logits, dim=1)[:, 0].cpu().numpy()  # Class 0 = FAKE

    # Apply confident strategy
    aggregated_prob = confident_strategy(probs, t=confident_t)

    # Video-level prediction
    video_prediction = 'fake' if aggregated_prob >= threshold else 'real'

    # Statistics
    total_frames = len(probs)
    avg_fake_prob = float(np.mean(probs))
    max_fake_prob = float(np.max(probs))
    min_fake_prob = float(np.min(probs))
    frames_above_threshold = np.sum(probs >= threshold)
    fake_ratio = frames_above_threshold / total_frames

    # Check which strategy was used
    sz = len(probs)
    fakes = np.count_nonzero(probs > confident_t)
    low_conf = np.count_nonzero(probs < 0.2)

    if fakes > sz // 2.5 and fakes > 11:
        strategy_used = f"High-confidence (>{confident_t}): {fakes} frames"
    elif low_conf > 0.9 * sz:
        strategy_used = f"Low-confidence (<0.2): {low_conf} frames"
    else:
        strategy_used = "Simple mean (all frames)"

    is_correct = (video_prediction == ground_truth)

    return {
        'video_name': video_name,
        'ground_truth': ground_truth,
        'prediction': video_prediction,
        'is_correct': is_correct,
        'total_frames': total_frames,
        'aggregated_prob': float(aggregated_prob),  # From confident strategy
        'avg_fake_prob': avg_fake_prob,  # Simple mean (for comparison)
        'max_fake_prob': max_fake_prob,
        'min_fake_prob': min_fake_prob,
        'frames_above_threshold': int(frames_above_threshold),
        'fake_ratio': fake_ratio,
        'strategy_used': strategy_used,
        'frame_probabilities': probs.tolist()
    }

# ============================================================================
# Test Confident Strategy on All Videos
# ============================================================================
print("="*80)
print("TESTING DFDC WINNER'S CONFIDENT STRATEGY")
print("="*80)
print("This uses high-confidence frames instead of simple averaging")
print("="*80)

# Get all video names
frames_base = Path('/content/drive/MyDrive/DeepfakeData/CompleteData/organized/crops/test')
all_video_names = set()

fake_dir = frames_base / 'fake_videos'
if fake_dir.exists():
    for frame_file in fake_dir.glob('*.jpg'):
        video_name = frame_file.stem.rsplit('_', 1)[0]
        all_video_names.add(video_name)

real_dir = frames_base / 'real_videos'
if real_dir.exists():
    for frame_file in real_dir.glob('*.jpg'):
        video_name = frame_file.stem.rsplit('_', 1)[0]
        all_video_names.add(video_name)

all_video_names = sorted(list(all_video_names))

print(f"Testing {len(all_video_names)} videos with confident strategy...")
print("="*80)

# Test with confident strategy
confident_results = []
for video_name in all_video_names:
    result = predict_video_with_confident_strategy(video_name, threshold=0.85, confident_t=0.8)
    if 'error' not in result:
        confident_results.append(result)

# Separate by ground truth
fake_videos_conf = [r for r in confident_results if r['ground_truth'] == 'fake']
real_videos_conf = [r for r in confident_results if r['ground_truth'] == 'real']

# Calculate metrics
fake_correct_conf = sum(1 for r in fake_videos_conf if r['is_correct'])
fake_total_conf = len(fake_videos_conf)
fake_accuracy_conf = fake_correct_conf / fake_total_conf if fake_total_conf > 0 else 0

real_correct_conf = sum(1 for r in real_videos_conf if r['is_correct'])
real_total_conf = len(real_videos_conf)
real_accuracy_conf = real_correct_conf / real_total_conf if real_total_conf > 0 else 0

total_correct_conf = fake_correct_conf + real_correct_conf
total_videos_conf = len(confident_results)
overall_accuracy_conf = total_correct_conf / total_videos_conf if total_videos_conf > 0 else 0

# Print results
print(f"\n{'='*80}")
print("RESULTS WITH CONFIDENT STRATEGY (threshold=0.85, confident_t=0.8)")
print(f"{'='*80}")
print(f"\nTotal Videos: {total_videos_conf}")
print(f"  Fake videos: {fake_total_conf}")
print(f"  Real videos: {real_total_conf}")

print(f"\nFAKE Videos:")
print(f"  Correctly identified as FAKE: {fake_correct_conf}/{fake_total_conf}")
print(f"  Accuracy: {fake_accuracy_conf:.4f} ({fake_accuracy_conf*100:.2f}%)")

print(f"\nREAL Videos:")
print(f"  Correctly identified as REAL: {real_correct_conf}/{real_total_conf}")
print(f"  Accuracy: {real_accuracy_conf:.4f} ({real_accuracy_conf*100:.2f}%)")

print(f"\nOVERALL:")
print(f"  Correct predictions: {total_correct_conf}/{total_videos_conf}")
print(f"  Overall accuracy: {overall_accuracy_conf:.4f} ({overall_accuracy_conf*100:.2f}%)")

# Compare with previous results (simple majority vote)
print(f"\n{'='*80}")
print("COMPARISON: Confident Strategy vs Simple Majority Vote")
print(f"{'='*80}")
print(f"{'Method':<30} {'Fake Acc':<12} {'Real Acc':<12} {'Overall Acc':<12}")
print("-"*70)
print(f"{'Simple Majority Vote':<30} {0.9697:<12.4f} {0.8750:<12.4f} {0.9512:<12.4f}")
print(f"{'Confident Strategy':<30} {fake_accuracy_conf:<12.4f} {real_accuracy_conf:<12.4f} {overall_accuracy_conf:<12.4f}")

improvement = overall_accuracy_conf - 0.9512
if improvement > 0:
    print(f"\n✓ Improvement: +{improvement:.4f} ({improvement*100:.2f}%)")
elif improvement < 0:
    print(f"\n✗ Decrease: {improvement:.4f} ({improvement*100:.2f}%)")
else:
    print(f"\n= No change")

# Show detailed results for failed videos (if any)
failed_videos_conf = [r for r in confident_results if not r['is_correct']]
if failed_videos_conf:
    print(f"\n{'='*80}")
    print(f"FAILED VIDEOS WITH CONFIDENT STRATEGY ({len(failed_videos_conf)}):")
    print(f"{'='*80}")
    for r in failed_videos_conf:
        print(f"  {r['video_name']}: GT={r['ground_truth']}, Pred={r['prediction']}")
        print(f"    Aggregated prob: {r['aggregated_prob']:.4f} (vs simple mean: {r['avg_fake_prob']:.4f})")
        print(f"    Strategy: {r['strategy_used']}")

# Show which strategy was used for each video
print(f"\n{'='*80}")
print("STRATEGY USAGE BREAKDOWN:")
print(f"{'='*80}")
strategy_counts = {}
for r in confident_results:
    strategy = r['strategy_used']
    if strategy not in strategy_counts:
        strategy_counts[strategy] = 0
    strategy_counts[strategy] += 1

for strategy, count in sorted(strategy_counts.items()):
    print(f"  {strategy}: {count} videos")

print("\n" + "="*80)
print("Analysis complete!")
print("="*80)


TESTING DFDC WINNER'S CONFIDENT STRATEGY
This uses high-confidence frames instead of simple averaging
Testing 134 videos with confident strategy...

RESULTS WITH CONFIDENT STRATEGY (threshold=0.85, confident_t=0.8)

Total Videos: 134
  Fake videos: 125
  Real videos: 9

FAKE Videos:
  Correctly identified as FAKE: 125/125
  Accuracy: 1.0000 (100.00%)

REAL Videos:
  Correctly identified as REAL: 0/9
  Accuracy: 0.0000 (0.00%)

OVERALL:
  Correct predictions: 125/134
  Overall accuracy: 0.9328 (93.28%)

COMPARISON: Confident Strategy vs Simple Majority Vote
Method                         Fake Acc     Real Acc     Overall Acc 
----------------------------------------------------------------------
Simple Majority Vote           0.9697       0.8750       0.9512      
Confident Strategy             1.0000       0.0000       0.9328      

✗ Decrease: -0.0184 (-1.84%)

FAILED VIDEOS WITH CONFIDENT STRATEGY (9):
  apvzjkvnwn: GT=real, Pred=fake
    Aggregated prob: 0.9944 (vs simple mean: 0.99